tardis/io/logger/logger.py

```diff
@@ -1,165 +1,265 @@
-
-        if not specific_log_level:
-            tardis_config["debug"][
-                "specific_log_level"
-            ] = DEFAULT_SPECIFIC_STATE
-            specific_log_level = tardis_config["debug"]["specific_log_level"]
-
-    logging_level = logging_level.upper()
-    if logging_level not in LOGGING_LEVELS:
-        raise ValueError(
-            f"Passed Value for log_level = {logging_level} is Invalid. Must be one of the following {list(LOGGING_LEVELS.keys())}"
+            for filter in logger.filters:
+                for logger in tardis_loggers:
+                    logger.removeFilter(filter)
+
+
+    def setup_widget_logging(self):
+        """Set up widget-based logging interface."""
+        # widget_handler = LoggingHandler(log_outputs, self.config.COLORS)
+        widget_handler = AsyncEmitLogHandler(log_outputs, self.config.COLORS)
+        widget_handler.setFormatter(
+            logging.Formatter("%(name)s [%(levelname)s] %(message)s (%(filename)s:%(lineno)d)")
         )
+        
+        self._configure_handlers(widget_handler)
+    
+    def _configure_handlers(self, widget_handler):
+        """Configure logging handlers."""
+        logging.captureWarnings(True)
+        self.logger.setLevel(logging.DEBUG)
 
-    # Getting the TARDIS logger & all its children loggers
-    logger = logging.getLogger("tardis")
+        for logger in [self.logger, logging.getLogger()]:
+            for handler in logger.handlers[:]:
+                logger.removeHandler(handler)
+        
+        self.logger.addHandler(widget_handler)
+        logging.getLogger("py.warnings").addHandler(widget_handler)
+        
 
-    # Creating a list for Storing all the Loggers which are derived from TARDIS
-    tardis_loggers = tardis_logger()
+def get_environment():
+    """Determine the execution environment"""
+    try:
+        import IPython
+        ipython = IPython.get_ipython()
+        
+        if ipython is None:
+            return 'standard'
+            
+        # Check for VSCode specific environment variables
+        if any(x for x in ('VSCODE_PID', 'VSCODE') if x in os.environ):
+            return 'vscode'
+            
+        # Check if running in Jupyter notebook
+        if 'IPKernelApp' in ipython.config:
+            return 'jupyter'
+            
+        return 'standard'
+    except:
+        return 'standard'
 
-    if logging_level in LOGGING_LEVELS:
-        for logger in tardis_loggers:
-            logger.setLevel(LOGGING_LEVELS[logging_level])
+class AsyncEmitLogHandler(logging.Handler):
+    def __init__(self, log_outputs, colors):
+        super().__init__()
+        self.log_outputs = log_outputs
+        self.colors = colors
+        self.environment = get_environment()
+        self.main_thread_id = threading.get_ident()
+        self.futures = []
+        
+        # Single event loop and thread for all environments
+        self.loop = asyncio.new_event_loop()
```

Comment:
Why do we need this to be created for environments outside of a notebook? e.g. what happens if you run TARDIS in the command line?

Reply:
I'm afraid it creates a new loop there too. Do we want to show the logs in the terminal or do we want to show it as a Panel widget in the browser(it would be deployed on some port)? If it is the first case then I will change it so that this mechanism only works with VSCode and Jupyter Notebooks(since those are the places where we would be seeing the widget). Does that sound alright to you?

Reply:
Yes, please make the logs just print to terminal for CLI use of TARDIS.